### PR TITLE
Create API for TopN functions

### PR DESCRIPTION
**Pull request by jbcrail**
This PR creates the Kibana server API so that the UI can eventually query the Top N functions.

The new API is accessible at `/api/prodfiler/v2/topn/functions?index={string}&projectID={number}&timeFrom={number}&timeTo={number}&startIndex={number}&endIndex={number}`. The UI does not use the API at the moment, so manual querying is required.

The implementation uses the same Elasticsearch queries as the flamegraph API. However, the only difference is how query results are packaged together in the JSON response.

Beyond the necessary work to implement the API, I also made a few minor refactors:
- Shared query for stacktrace events across flamegraph and TopN functions
- Shared stacktrace fixtures across tests
- Removed unused `parallelMget` function

Related to https://github.com/elastic/prodfiler/issues/2363

### CODE REVIEW

File: src/plugins/profiling/common/functions.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      }
+
+      const topNFunction = topNFunctions.get(frameGroupID)!;
+
+      if (!uniqueFrameGroupsPerEvent.has(frameGroupID)) {
+        uniqueFrameGroupsPerEvent.add(frameGroupID);
+        topNFunction.CountInclusive += count;
+      }
+
+      if (i === 0) {
```
Comment (rockdaboot) [inline src/plugins/profiling/common/functions.ts]: We reversed the order of frames in the backend. Do we reverse it again when parsing the query results ? (If not, this should be `if (i === frames.length - 1)`.)

File: src/plugins/profiling/server/routes/functions.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  mgetStackFrames,
+  mgetStackTraces,
+  searchEventsGroupByStackTrace,
+  searchStackTraces,
+} from './stacktrace';
+import { getClient } from './compat';
+
+async function queryTopNFunctions(
+  logger: Logger,
+  client: ElasticsearchClient,
```
Comment (rockdaboot) [inline src/plugins/profiling/server/routes/functions.ts:27]: Using the same names as at other places, like `esClient` and `targetSampleSize` makes it easier to follow, IMO.
There are other places as well where we have different names for the same variable/value. When jumping around in the code, it makes it easier to recognize/follow the code with consistent names.
We don't have to do the necessary changes in this PR, but if not, can you open a 'cleanup' issue ?

**Review by rockdaboot — APPROVED** [review]
:+1: (Just have some minor comments / questions)

File: src/plugins/profiling/common/functions.ts (hunk 1 above)
Comment (jbcrail) [reply src/plugins/profiling/common/functions.ts]: Thanks, the stack frames used to build the TopN functions response was using the leaf-frame-last ordering so I fixed it as you suggested.

File: src/plugins/profiling/server/routes/functions.ts (hunk 2 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts:27]: As I was working on other features, I also noticed naming inconsistencies, opportunities to reduce duplication, linting warnings, and unidiomatic TypeScript. I've been cleaning this up incrementally as I go along, but making the work explicit by creating a cleanup issue makes sense as we get closer to MVP and prepare to merge into the main Kibana repo.

File: src/plugins/profiling/server/routes/stacktrace.ts [hunk 3]
```diff
@@ -166,6 +168,80 @@ export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
   } as StackTrace;
 }
 
+export async function searchEventsGroupByStackTrace(
+  logger: Logger,
+  client: ElasticsearchClient,
+  index: DownsampledEventsIndex,
+  filter: ProjectTimeQuery
+) {
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/stacktrace.ts:176]: Generally we try to avoid more than 2/3 arguments. Once we reach that threshold we move to a single argument with named properties.

File: src/plugins/profiling/server/routes/stacktrace.ts [hunk 4]
```diff
@@ -166,6 +168,80 @@ export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
...
+  // Using filter_path is less readable and scrollSearch seems to be buggy - it
+  // applies filter_path only to the first array of results, but not on the following arrays.
+  // The downside of `_source` is: it takes 2.5x more time on the ES side (see "took" field).
+  // The `composite` keyword skips sorting the buckets as and return results 'as is'.
+  // A max bucket size of 100000 needs a cluster level setting "search.max_buckets: 100000".
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/stacktrace.ts]: What is this referring to? Is it from copying/pasting?

File: src/plugins/profiling/server/routes/stacktrace.ts [hunk 5]
```diff
@@ -166,6 +168,80 @@ export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
...
+      return await client.search(
+        {
+          index: index.name,
+          track_total_hits: false,
+          query: filter,
+          aggs: {
+            group_by: {
+              terms: {
+                // 'size' should be max 100k, but might be slightly more. Better be on the safe side.
+                size: 150000,
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/stacktrace.ts:190]: The default cluster setting for max buckets is 64k I believe. Do we expect the default cluster setting to be updated at some point?

File: src/plugins/profiling/server/routes/stacktrace.ts [hunk 6]
```diff
@@ -166,6 +168,80 @@ export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
...
+          },
+        }
+      );
+    }
+  );
+
+  const totalCount: number = getAggs(resEvents)?.total_count.value;
+  const stackTraceEvents = new Map<StackTraceID, number>();
+
+  await logExecutionLatency(logger, 'processing events data', async () => {
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/stacktrace.ts]: Is this actually doing anything asynchronous?

File: src/plugins/profiling/server/routes/functions.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        query: querySchema,
+      },
+    },
+    async (context, request, response) => {
+      try {
+        const { index, projectID, timeFrom, timeTo, startIndex, endIndex }: QuerySchemaType =
+          request.query;
+        const targetSampleSize = 20000; // minimum number of samples to get statistically sound results
+        const esClient = await getClient(context);
+        const filter = newProjectTimeQuery(projectID, timeFrom, timeTo);
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts]: I think more idiomatic naming would be `buildProjectTimeQuery` or `createProjectTimeQuery`. I rarely see `new` (except for the keyword).

File: src/plugins/profiling/server/routes/functions.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+          endIndex,
+          targetSampleSize
+        );
+        logger.info('returning payload response to client');
+
+        return response.ok({
+          body: topNFunctions,
+        });
+      } catch (e) {
+        logger.error('Caught exception when fetching Top N functions data: ' + e.message);
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts]: I think it's useful here to log the full error object, rather than just the message, which loses the stack trace.

File: src/plugins/profiling/server/routes/functions.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  index: string,
+  filter: ProjectTimeQuery,
+  startIndex: number,
+  endIndex: number,
+  sampleSize: number
+): Promise<any> {
+  const testing = index === 'profiling-events2';
+  if (testing) {
+    index = 'profiling-events-all';
+  }
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts]: Maybe this should be moved out of this function? Is it still being used?

File: src/plugins/profiling/server/routes/functions.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    filter
+  );
+
+  // Manual downsampling if totalCount exceeds sampleSize by 10%.
+  if (totalCount > sampleSize * 1.1) {
+    const p = sampleSize / totalCount;
+    logger.info('downsampling events with p=' + p);
+    await logExecutionLatency(logger, 'downsampling events', async () => {
+      totalCount = downsampleEventsRandomly(stackTraceEvents, p, filter.toString());
+    });
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts:55]: This is also not doing anything asynchronous, presumably?

Comment (dgieselaar) [reply src/plugins/profiling/server/routes/functions.ts:55]: I think this has been discussed before, but if we are using random sampling, why not use the random_sampler aggregation?

File: src/plugins/profiling/server/routes/functions.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    logger.info('downsampled total count: ' + totalCount);
+    logger.info('unique downsampled stacktraces: ' + stackTraceEvents.size);
+  }
+
+  // profiling-stacktraces is configured with 16 shards
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts]: is this comment still relevant?

File: src/plugins/profiling/server/routes/functions.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  const { stackTraces, stackFrameDocIDs, executableDocIDs } = testing
```
Comment (dgieselaar) [inline src/plugins/profiling/server/routes/functions.ts]: do we still need both branches? (where testing is either true or false)

File: src/plugins/profiling/common/functions.test.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+import { createTopNFunctions } from './functions';
+
+import { events, stackTraces, stackFrames, executables } from './__fixtures__/stacktraces';
+
+describe('TopN function operations', () => {
+  test('1', () => {
+    const topNFunctions = createTopNFunctions(events, stackTraces, stackFrames, executables, 0, 5);
+
+    expect(topNFunctions.TotalCount).toEqual(40);
```
Comment (dgieselaar) [inline src/plugins/profiling/common/functions.test.ts:17]: What we do in the APM UI is avoid using value testing. Usually we generate synthetic data with some parameters and expect those parameters to be reflected in the results from our functions/APIs. It's helpful in making our tests more understandable and maintainable.

File: src/plugins/profiling/common/__fixtures__/stacktraces.ts [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+const defaultStackFrame = {
+  FileName: '',
+  FunctionName: '',
+  FunctionOffset: 0,
+  LineNumber: 0,
+  SourceType: 0,
+};
+
+export const stackFrames = new Map([
```
Comment (dgieselaar) [inline src/plugins/profiling/common/__fixtures__/stacktraces.ts:120]: I see a lot of Map/Set usage, why's that? I feel like they don't really add a lot of value over plain objects/arrays, only in some cases where you use things like generators.

File: src/plugins/profiling/common/functions.ts [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+  const topNFunctions = new Map<FrameGroupID, TopNFunctionAndFrameGroup>();
+
+  // Collect metadata and inclusive + exclusive counts for each distinct frame.
+  for (const [traceHash, count] of events) {
+    const uniqueFrameGroupsPerEvent = new Set<FrameGroupID>();
+
+    totalCount += count;
+
+    const frames = metadata.get(traceHash)!;
+    for (let i = 0; i < frames.length; i++) {
```
Comment (dgieselaar) [inline src/plugins/profiling/common/functions.ts:64]: Why use a for loop here, and forEach in other places? I don't really care, just wondering if it's worth picking one or the other for the purpose of consistency.

File: src/plugins/profiling/common/functions.ts [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+      const frameGroup = defaultGroupBy(frames[i]);
+      const frameGroupID = hashFrameGroup(frameGroup);
+
+      if (!topNFunctions.has(frameGroupID)) {
+        topNFunctions.set(frameGroupID, {
+          Frame: frames[i],
+          FrameGroup: frameGroup,
+          CountExclusive: 0,
+          CountInclusive: 0,
+        } as TopNFunctionAndFrameGroup);
```
Comment (dgieselaar) [inline src/plugins/profiling/common/functions.ts]: is the type assertion here necessary?

**Review by dgieselaar — APPROVED** [review]
I've opted for a little bit of nitpicking to get a better understanding of what kind of feedback you are looking for, so feel free to ignore whatever you feel is not applicable. Generally it looks in good shape!

File: src/plugins/profiling/server/routes/functions.ts (hunk 12 above)
Comment (rockdaboot) [reply src/plugins/profiling/server/routes/functions.ts]: No, we are mostly done with it, I'd say.

File: src/plugins/profiling/server/routes/stacktrace.ts (hunk 4 above)
Comment (rockdaboot) [reply src/plugins/profiling/server/routes/stacktrace.ts]: This is indeed a remnant from times where we use the composite aggregation for pagination.

File: src/plugins/profiling/server/routes/stacktrace.ts (hunk 5 above)
Comment (rockdaboot) [reply src/plugins/profiling/server/routes/stacktrace.ts:190]: We still have to get rid of this, but currently have no good concept of pagination without the scroll API (which is deprecated) or composite aggregation (which didn't combine with terms aggregation, AFAIR). This is beyond this PR (the code has just been shifted around), but any help is appreciated (we postponed this so far).

File: src/plugins/profiling/common/functions.ts (hunk 15 above)
Comment (rockdaboot) [reply src/plugins/profiling/common/functions.ts:64]: In this case we need a special handling for the last element further down.
And in general I personally prefer explicit for loops in Typescript and forEach in JS.
What is the general consent in Elastic about this ? Is there a style guide or recommendations ?

Comment (dgieselaar) [reply src/plugins/profiling/common/functions.ts:64]: I think `forEach` is more common. There's an explicit lint rule against `for(... in ...)` too.

Comment (rockdaboot) [reply src/plugins/profiling/common/functions.ts:64]: Since I work most of the time with Go, C/C++, Java, the common denominator is a for loop.
Indeed, for me it is much easier to read an explicit for loop than a forEach construction.
And therefore I like Google's recommendations for idiomatic Typescript: https://google.github.io/styleguide/tsguide.html#iterating-containers
😃

File: src/plugins/profiling/common/functions.ts (hunk 16 above)
Comment (jbcrail) [reply src/plugins/profiling/common/functions.ts]: No, it's redundant since the container is typed already. I also removed another type assertion, thanks!

File: src/plugins/profiling/server/routes/functions.ts (hunk 7 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts]: Agreed, I renamed it to `createProjectTimeQuery` since it fits the naming convention we've used more often.

File: src/plugins/profiling/server/routes/functions.ts (hunk 8 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts]: Thanks, we're now logging the full object everywhere.

File: src/plugins/profiling/common/functions.test.ts (hunk 13 above)
Comment (jbcrail) [reply src/plugins/profiling/common/functions.test.ts:17]: Are there areas in the Kibana codebase that are good examples of this? I'd love to do this going forward.

File: src/plugins/profiling/common/__fixtures__/stacktraces.ts (hunk 14 above)
Comment (jbcrail) [reply src/plugins/profiling/common/__fixtures__/stacktraces.ts:120]: Good question:

1. We use `Map`/`Set`s over plain objects/arrays since the former have similar semantics to Go's maps (non-strings as keys, built-in retrieval of map/set size, etc). Most of the code was ported from Go and relies on those semantic assumptions.
2. We often de-duplicate the data before sending down the UI so `Set`s make our intent more explicit than plain objects.
3. This is more of a personal preference, but with `Map`s I can ignore the historical edge cases when using plain objects as maps.

I'm still somewhat new to TypeScript so I probably haven't picked up all of the idioms and conventions.

File: src/plugins/profiling/common/functions.test.ts (hunk 13 above)
Comment (dgieselaar) [reply src/plugins/profiling/common/functions.test.ts:17]: We use our own tool called `synthtrace` in APM API tests. It's meant for generating timeseries data. Here's an example of generating both the data and running the assertions: https://github.com/elastic/kibana/blob/7af6915581f0733d3dd82c283d1cc0f67a939e12/x-pack/test/apm_api_integration/tests/latency/service_apis.spec.ts

File: src/plugins/profiling/server/routes/functions.ts (hunk 9 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts]: I removed this since we're no longer using it as discussed below.

File: src/plugins/profiling/server/routes/functions.ts (hunk 11 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts]: No, I removed it, thanks.

File: src/plugins/profiling/server/routes/functions.ts (hunk 12 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts]: I removed it, thanks.

File: src/plugins/profiling/server/routes/stacktrace.ts (hunk 4 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/stacktrace.ts]: Thanks, I removed it.

File: src/plugins/profiling/server/routes/stacktrace.ts (hunk 6 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/stacktrace.ts]: No, I think we need this boilerplate to log the execution time. However, the relevant code is just assigning to a map so logging the time isn't interesting enough for the overall latency, thus, I removed it.

File: src/plugins/profiling/server/routes/functions.ts (hunk 10 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/functions.ts:55]: I think @rockdaboot or @inge4pres have the fuller story on this. If any action is needed, we do it in another PR.

File: src/plugins/profiling/server/routes/stacktrace.ts (hunk 3 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/stacktrace.ts:176]: Thanks, I moved this into https://github.com/elastic/prodfiler/issues/2399 since this will require additional work.

File: src/plugins/profiling/common/functions.test.ts (hunk 13 above)
Comment (jbcrail) [reply src/plugins/profiling/common/functions.test.ts:17]: Thanks, I spawned this off into https://github.com/elastic/prodfiler/issues/2411.

**Comment by jbcrail** [comment]
Thanks for everyone's feedback. For those items that I didn't address, I created issues so that we can continue the work later.

I'll merge this PR now so we can move to the new repo.
